src/js/cursor.ts

```diff
@@ -0,0 +1,159 @@
+import { KEYS } from "./constants.js";
+import { createCursorSpan } from "./utils.js";
+
+export class Cursor {
+  position: number;
+  element: HTMLSpanElement;
+
+  constructor() {
+    this.position = 0;
+    this.element = document.getElementById("cursor") as HTMLSpanElement;
+  }
+
+  private remove() {
+    document.getElementById("cursor")!.remove();
+  }
+
+  private place(input: null | HTMLSpanElement) {
+    let content = input?.textContent;
+    this.element = createCursorSpan();
+    input!.innerHTML =
+      content?.slice(0, this.position) +
+      this.element.outerHTML +
+      content?.slice(this.position);
+  }
+
+  resetPosition() {
+    this.position = 0;
+  }
+
+  setPosition(position: number, input: null | HTMLSpanElement) {
+    if (position == 0) {
+      this.position = 0;
+      return;
+    }
+    if (position > 0 && position <= input!.textContent!.length) {
+      this.position = position;
+    }
+  }
+
+  addTextContent(input: null | HTMLSpanElement, textContent: string) {
+    this.remove();
+    let content = input!.textContent;
+    input!.textContent =
+      content!.slice(0, this.position) +
+      textContent +
+      content!.slice(this.position);
+    this.position++;
+    this.place(input);
+  }
+
+  move(
+    input: null | HTMLSpanElement,
+    key: string,
+    isAltPressed: boolean,
+    isMetaPressed: boolean
+  ) {
+    if (!input?.textContent || !input.textContent.length) {
+      return;
+    }
+    if (key === KEYS.ARROW_LEFT && this.position > 0) {
+      this.remove();
+      if (isAltPressed) {
+        this.setPosition(
+          this.getAltIndexLeftFromCursor(input?.textContent),
+          input
+        );
+      } else if (isMetaPressed) {
+        this.resetPosition();
+      } else {
+        this.setPosition(this.position - 1, input);
+      }
+      this.place(input);
+    }
+
+    if (
+      key === KEYS.ARROW_RIGHT &&
+      this.position < input.textContent.length - 1 // here we are subtracting 1 because of cursor character
+    ) {
+      this.remove();
```

Comment:
For `remove` and `place` we should consider using [decorators](https://www.geeksforgeeks.org/what-are-decorators-and-how-are-they-used-in-javascript/).